doc/manual/man/man1/convert_trace.1
.TH "convert_trace" 1 "" "" "Staden Package"
.SH "NAME"
.PP
convert_trace \- Converts trace file formats

.SH "SYNOPSIS"
.PP

\fBconvert_trace\fP
[\fB-in_format\fP \fIformat\fP]
[\fB-out_format\fP \fIformat\fP]
[\fB-fofn\fP \fIfile_of_filenames\fP]
[\fB-passed\fP \fIfofn\fP]
[\fB-failed\fP \fIfofn\fP]
[\fB-name\fP \fIid\fP]
[\fB-subtract_background\fP]
[\fB-normalise\fP]
[\fB-scale\fP \fIrange\fP]
[\fB-compress\fP \fImode\fP]
[\fB-abi_data\fP \fIcounts\fP]
[\fIinformat\fP \fIoutformat\fP]

.SH "DESCRIPTION"
.PP

\fBconvert_trace\fP converts between the various DNA sequence chromatogram
formats, optionally performing trace processing actions too. It can read ABI
(raw or processed), ALF, CTF, SCF and ZTR formats. It can write CTF, EXP, PLN, 
SCF and ZTR formats. (Note that EXP (Experiment File) and PLN formats are
text sequences rather than a binary trace.)

There are two main modes of operation; either with a file of filenames
specified using the \fB-fofn\fP \fIfilename\fP option, or acting as a filter
to process one single file. In this case the input and output file format may
be specified as the last two options on the command line.

.SH "OPTIONS"
.PP
.TP
\fB-abi_data\fP \fIcounts\fP
Only of use when processing ABI files. This indicates which ABI
\fBDATA\fP channel numbers to use. For sequencing files this defaults to
"9,10,11,12" which corresponds to the processed data. To read the raw data 
use "1,2,3,4".

.TP
\fB-compress\fP \fImode\fP
Specifies the name of a program to use to compress the trace data prior to 
writing. Due to limitations in the current implementation this option does 
not work when \fBconvert_trace\fP is operating as a filter (and so
requires use of the \fB-fofn\fP option). Valid values for \fImode\fP are
compress, bzip, bzip2, gzip, pack and szip. Note that for ZTR, ZTR2 and
ZTR3 format files specifying compression modes will not reduce the file
size as this format already contains internal compression algorithms. The
ZTR1 format does not internally compress and so \fB-compress\fP will have 
an effect.

.TP
\fB-failed\fP \fIfofn\fP
Produces a file listing the filenames which have failed to be
converted. This only makes sense when also using \fB-fofn\fP.

.TP
\fB-fofn\fP \fIfile_of_filenames\fP
Processes several files instead of one, with the filenames to read from and
written to being listed in \fIfile_of_filenames\fP with one pair (input and
output filenames) being listed per line, separated by spaces. If the
filenames contain spaces then these may be "escaped" using
backslashes. Similarly backslashes should be escaped using a double
backslash. For example to convert "file a.scf" and "fileb.scf" to "file
a.ztr" and "fileb.ztr" respectively we would use a \fIfile_of_filenames\fP
containing:

.nf
.in +0.5i
file\\ a.scf    file\\ a.ztr
fileb.scf      fileb.ztr
.in -0.5i
.fi

.TP
\fB-in_format\fP \fIformat\fP
Specifies the format for the input data. Typically the input format is
automatically determined so this may not be required. \fIformat\fP should be 
one of ABI, ALF, CTF, EXP, PLN, SCF, ZTR, ZTR1, ZTR2 or ZTR3. The ZTR
formats all conform to the ZTR specification, but this indicates the
compression level to be used.

.TP
\fB-name\fP \fIid\fP
When producing an Experiment File this specifies the value of the
\fBID\fP line. Without this option default Experiment File ID line is the 
output filename, or if this is stdout it is the input filename.

.TP
\fB-normalise\fP
Attempts to normalise the trace amplitudes to produce more even height
peaks. This may be useful to compensate for large spikes at either the
start or end of the trace.

.TP
\fB-out_format\fP \fIformat\fP
Specifies the output format for all files, whether read from a file of
filenames or via a filter.  \fIformat\fP should be 
one of ABI, ALF, CTF, EXP, PLN, SCF, ZTR, ZTR1, ZTR2 or ZTR3. The ZTR
formats all conform to the ZTR specification, but this indicates the
compression level to be used.

.TP
\fB-passed\fP \fIfofn\fP
Produces a file listing the filenames which have been successfully
converted. This only makes sense when also using \fB-fofn\fP.

.TP
\fB-scale\fP \fIrange\fP
Scales all trace amplitudes so that they fit within the range of 0 to 
\fIrange\fP inclusive. Any integer value of \fIrange\fP may be used between 1
and 65535, but this option is designed for down-scaling traces in order to 
reduce file size.

.TP
\fB-subtract_background\fP
Attempts to remove background trace levels by analysing each trace channel 
independently to determine the baseline. This option is mainly used when
processing raw data.
.TE
.SH "EXAMPLES"
.PP

To convert several files to ZTR format using the same example file of
filenames listed in the \fB-fofn\fP option above:

.nf
.in +0.5i
convert_trace -out_format ZTR -fofn filename
.in -0.5i
.fi

To subtract the background from a raw ABI file and save this as an SCF file:

.nf
.in +0.5i
convert_trace -abi_data 1,2,3,4 -subtract_background ABI SCF < a.abi > a.scf
.in -0.5i
.fi

.SH "NOTES"
.PP

If ABI files are manually edited before input to convert_trace then the
internal formats of these files may differ to the format expected by
convert_trace.

.SH "SEE ALSO"
.PP

\fBscf\fR(4)
\fBztr\fR(4)
\fBmakeSCF\fR(1)



NEWS
-----------------------------------------------------------------------------
30th September 2005
===================

Version 1.6.0 tagged

-----------------------------------------------------------------------------

27th September 2005
===================

Gap4
----

* Renamed the discrepancy plot to be second highest confidence plot.

* Fixed some (apparently rare) find internal joins crashes.

* Added a minimum quality slider to the Report Mutations "by
  differences" mode. Also fixed crashes and missing pictures in the
  HTML report.

io_lib
------

* Changed the HashFile format and updated it to be official version
  1.00.

  The key difference is that there's now a footer pointing back into
  the hashfile header allowing for the index to simply be appended to
  a (e.g. tar) archive.

* Improved handling of 454 data in ZTR; it no longer attempts to
  compress raw pyrosequencing data as this was not working well.

* Fixed Bug ID 1289095 (file descriptor leak).

-----------------------------------------------------------------------------

24th August 2005
================

Gap4
----

* Rewrote the Shuffle Pads algorithm to be more like ReAligner. It can
  introduce new pads and remove old ones as necessary in order to
  optimise the alignments. It has also moved from inside the contig
  editor to part of the main menu so that it can be scripted easier.

* The join editor invoked from Find Read Pairs plot now brings up the
  editors at the start/end of the sequences forming the read-pair
  instead of the start/end of the two contigs.

* Search by sequence in the editor now finds consensus matches
  too. Also fixed some bugs where cutoff data matches were not always
  found and in some cases (reverse search) non-cutoff matches could be
  missed.

* The SNP Candidates plot now allows for the consensus to be saved
  directly from this window. The consensus and fake sequences added
  when a real split is requested now includes the "unsplit" readings
  from regions which apparently contained no SNPs (and therefore
  belong to all split contigs).

* Report Mutations bug fix. Producing HTML output now works (albeit
  with less content) when asking for SNPs to be derived from
  "differences" instead of tags.

Trev
----

* Pyrosequencing data (eg 454 flowgrams) are now supported. Trev
  auto-senses the type and views accordingly. Added format for 454's
  SFF standard too.

Io_lib
------

* New release - 1.9.0. No longer uses libtool in configuration.

* Added "HASH=hashfile" component of RAWDATA search path along with
  hash_tar and hash_extract tools. These operate like index_tar but
  provide true fast random access into tar files (or other formats if
  needed).

* Major update of I/O mechanisms to use in-memory equivalents of the
  files being read from or written to. This has lead to some
  incompatibilities (see src/io_lib/CHANGES), but is also an order of
  magnitude faster in many cases.

Spin
----

* Under MS Windows spin and EMBOSSwin should coexist happily to the
  extent that Spin will spot the EMBOSSwin installation and use it.

Pregap4
-------

* Fixed various space-in-pathname bugs (most prevalent on MS Windows).


Misc
----

* Removed some redundant old ABI utilities: getABIfloat, getABIhex,
  getABIraw, getABIstring. All of these can be achieved via getABIfield.


-----------------------------------------------------------------------------

3 May 2005
==========

Internal Sanger test version rel-1-6-0b3 ("gap4 -test")

Gap4
----

* Substantial modifications to the gap4 I/O mechanisms (balancing the
  "freetree"). The upshot of this is that the worst-case performance
  is now dramatically improved and the average performance will be
  slightly better too. Complementing a 10Mb contig dropped from about
  50 minutes to 1.5 minutes. Disassembling all readings (approx
  175000) in this database went from ~3 hours to 2.5 minutes.

* The contig editor "sequence sets" are now coloured to aid
  distinguishing.

* Fixed bugs with X11 BadDrawable crashes (I hope).

* Reverted the OLIGo tag back to using "serial=#" as the first line,
  although the additional information is still kept (albeit in a
  reduced precision format to make the tag shorter).

-----------------------------------------------------------------------------

29 March 2005
=============

Internal Sanger test version rel-1-6-0b2 (the previous "gap4 -test")

Gap4
----

* Added 64-bit filesize support, meaning that Gap4 databases can now
  be larger than 2Gb in file size. Specify "-bitsize 64" to force
  64-bit mode, otherwise databases will still be created in the old
  32-bit format. Copy_db has a -b flag to change bitsize when copying
  databases so this aids in migration.

* A new experimental option "SNP candidates" (in the main view
  menu). Primarily designed for pulling apart the two alleles from
  mixed shotguns, it also has uses for exploring and sometimes
  resolving collapsed repeats. User feedback at this stage is most
  welcome.

* The consistency plots now distinguish between the old "discrepancy
  plot" (now named diploid discrepancy plot) and a new discrepancy
  plot that takes does not have the "two alleles only in 50/50 ratios"
  detector. This makes it more suitable for finding potential
  collapsed repeats.

* The contig editor now has the notion of "sets". A set is basically a
  sub-group of a contig, which is useful for exploring the outcome
  within a conti of pulling apart a collapsed repeat without actually
  doing it. The editor is automatically invoked in this manner by the
  SNP candidates output. This is largely experimental, but there are a
  number of potential ways to improve interaction. Ideas on a postcard
  please!

* Lots of speedups (primarily for the huge 64-bit databases):
  - Drawing of some consistency graphs;
  - Experiment file reading on large files (1000 fold on a 1Mb sequence);
  - Dealing with lists of contigs (eg "all contigs") is now O(N)

src/io_lib/progs/convert_trace.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "Read.h"
#include "traceType.h"
#include "seqIOABI.h"
#include "open_trace_file.h"

#define MAX(a,b) ((a)>(b)?(a):(b))

struct opts {
    char *name;
    char *fofn;
    char *passed;
    char *failed;
    int in_format;
    int out_format;
    int scale;
    int sub_background;
    int subtract;
    int normalise;
    int min_normalise;
    int compress_mode;
    int dots;
};

/*
 * Scales trace values from 0 to scale, but only if they are larger.
 */
void rescale_trace(Read *r, int scale) {
    double s;
    int i;

    if (r->maxTraceVal <= scale)
	return;
    
    s = ((double)scale)/r->maxTraceVal;

    for (i = 0; i < r->NPoints; i++) {
	r->traceA[i] = r->traceA[i] * s + 0.5;
	r->traceC[i] = r->traceC[i] * s + 0.5;
	r->traceG[i] = r->traceG[i] * s + 0.5;
	r->traceT[i] = r->traceT[i] * s + 0.5;
    }

    r->maxTraceVal = scale;
}

#if 0
/* OLD method, treats all channels together and assumes the same baseline for
 * each
 */
/*
 * Here we just take the minimum trace value and subtract this from all others.
 * The assumption is that the signal will always be 'base line' on at least
 * one of the four channels.
 */
void subtract_background(Read *r) {
    int i, min;
    for (i = 0; i < r->NPoints; i++) {
	min = 999999;
	if (r->traceA[i] < min) min = r->traceA[i];
	if (r->traceC[i] < min) min = r->traceC[i];
	if (r->traceG[i] < min) min = r->traceG[i];
	if (r->traceT[i] < min) min = r->traceT[i];
	r->traceA[i] -= min;
	r->traceC[i] -= min;
	r->traceG[i] -= min;
	r->traceT[i] -= min;
    }
}
#endif

static void subtract_background_ch(TRACE *channel, int nchannel) {
    int i, j, bg;
    int win_len = 501, win_len2 = win_len/2;
    TRACE *copy;

    if (NULL == (copy = (TRACE *)malloc(sizeof(*copy) * nchannel)))
	return;

    if (nchannel < win_len)
	win_len = nchannel;

    /* Take lowest background over win_len and subtract it */
    for (i = 0; i < nchannel; i++) {
	/* Could optimise this considerably */
	bg = INT_MAX;
	for (j = -win_len2; j < win_len2; j++) {
	    if (i+j < 0) continue;
	    if (i+j >= nchannel) break;
		
	    if (channel[i + j] < bg)
		bg = channel[i + j];
	}
	
	copy[i] = channel[i] - bg;
    }

    memcpy(channel, copy, nchannel * sizeof(*copy));
    free(copy);
}

/*
 * Find the average background level of a trace, and subtract this from the
 * peak heights.
 */
void subtract_background(Read *r) {
    subtract_background_ch(r->traceA, r->NPoints);
    subtract_background_ch(r->traceC, r->NPoints);
    subtract_background_ch(r->traceG, r->NPoints);
    subtract_background_ch(r->traceT, r->NPoints);
}

int int_compar(const void *a, const void *b) {
    return *(const TRACE *)a - *(const TRACE *)b;
}

int find_bg(TRACE *data, int ndata) {
    int i, bg;
    TRACE *copy = (TRACE *)malloc(ndata * sizeof(TRACE));

    /* Sort the trace samples by amplitude */
    memcpy(copy, data, ndata * sizeof(TRACE));
    qsort(copy, ndata, sizeof(TRACE), int_compar);

    /* Find the first non-zero value */
    for (i = 0; i < ndata && !copy[i]; i++)
	;

    /*
     * Now take a slie 0.05 through the remainder of the array and set this
     * as our background.
     */
    bg = copy[(int)((ndata - i) * 0.05 + i)];

    free(copy);
    return bg;
}

void trace_freq(TRACE *data, int ndata) {
    int i, bg;
    bg = find_bg(data, ndata);

    for (i = 0; i < ndata; i++) {
	data[i] = MAX(data[i] - bg, 0);
    }
}

/*
 * Separates out the dyes using a deconvolution matrix.
 * The order of elements in the matrix is C A G T.
 * A test matrix for the 373. Taken from the BASS distribution.
 */
double matrix[5][4] = {
   0.002439782,        -0.0015053751,       0.00011857301,    2.8906948e-06,
  -0.00075353298,       0.0032971052,      -0.006198165,      0.00014828549,
   0.00020249287,      -0.0017620348,       0.010530438,     -0.0020235507,
  -0.001144423,        -4.857673e-06,      -0.0018845701,     0.00395431,
  -0.12451385,          0.368916,          -2.928292,        -3.3142638
};
void separate_dyes(Read *r, double M[][4]) {
    int i, j;

    for (i = 0; i < r->NPoints; i++) {
	int C, A, G, T;
	double sep[4];

	C = r->traceC[i];
	A = r->traceA[i];
	G = r->traceG[i];
	T = r->traceT[i];

	for (j = 0; j < 4; j++)
	  sep[j] = C*M[0][j] + A*M[1][j] + G*M[2][j] + T*M[3][j] + M[4][j];

	for (j = 0; j < 4; j++)
	    sep[j] += 10;

	/* hack!
	   sep[0] += 0.1;
	   sep[1] += -0.4;
	   sep[2] += 2.9;
	   sep[3] += 3.2;
	*/

	r->traceC[i] = sep[0] < 0 ? 0 : 1000 * sep[0];
	r->traceA[i] = sep[1] < 0 ? 0 : 1000 * sep[1];
	r->traceG[i] = sep[2] < 0 ? 0 : 1000 * sep[2];
	r->traceT[i] = sep[3] < 0 ? 0 : 1000 * sep[3];
    }
}

/*
 * Find the maximum height of traces at the called bases. Use this to clip any
 * other bases.
 */
void reset_max_called_height(Read *r) {
    int i, max = 0;

    /* Find max */
    for (i=0; i < r->NBases; i++) {
	switch(r->base[i]) {
	case 'a':
	case 'A':
	    if (r->traceA[r->basePos[i]] > max)
		max = r->traceA[r->basePos[i]];
	    break;

	case 'c':
	case 'C':
	    if (r->traceC[r->basePos[i]] > max)
		max = r->traceC[r->basePos[i]];
	    break;

	case 'g':
	case 'G':
	    if (r->traceG[r->basePos[i]] > max)
		max = r->traceG[r->basePos[i]];
	    break;

	case 't':
	case 'T':
	    if (r->traceT[r->basePos[i]] > max)
		max = r->traceT[r->basePos[i]];
	    break;
	}
    }

    /* Clip to max */
    for (i = 0; i < r->NPoints; i++) {
	if (r->traceA[i] > max)
	    r->traceA[i] = max;
	if (r->traceC[i] > max)
	    r->traceC[i] = max;
	if (r->traceG[i] > max)
	    r->traceG[i] = max;
	if (r->traceT[i] > max)
	    r->traceT[i] = max;
    }
    if (r->maxTraceVal > max)
	r->maxTraceVal = max;
}

/*
 * Rescales peak heights based on a moving "marker". The marker tracks
 * up and down (attack and decay) based on the difference between itself and
 * the trace envelope. We then divide by the marker value to attempt to
 * normalise peak heights.
 *
 * min_marker is used to avoid scaling up noise and represents the minimum
 * value the marker is allowed to reach. Make sure it is > 0 or divide by
 * zero may occur.
 */
void rescale_heights(Read *r, int min_marker) {
    double marker = 0;
    int i, j, max, mtv = 0;
    TRACE *tx[4];

    tx[0] = r->traceA;
    tx[1] = r->traceC;
    tx[2] = r->traceG;
    tx[3] = r->traceT;

    for (i = 0; i < r->NPoints; i++) {
	for (max = j = 0; j < 4; j++)
	    if (max < tx[j][i])
		max = tx[j][i];
	if (!marker) {
	    marker = max;
	} else {
	    if (max >= marker) {
		/* attack */
		marker += (max - marker) / 20.0;
	    } else {
		/* decay */
		marker -= (marker - max) / 10.0;
	    }
	}
	if (marker < min_marker)
	    marker = min_marker;

	for (j = 0; j < 4; j++) {
	    double new = tx[j][i] * 2000.0/marker;
	    tx[j][i] = new > 32767 ? 32767 : new;
	    if (mtv < tx[j][i])
		mtv = tx[j][i];
	}

    }

    r->maxTraceVal = mtv;
}


int convert(mFILE *infp, mFILE *outfp, char *infname, char *outfname,
	    struct opts *opts) {
    Read *r;

    if (NULL == (r = mfread_reading(infp, infname, opts->in_format))) {
	fprintf(stderr, "failed to read file %s\n", infname);
	return 1;
    }

    if (opts->subtract) {
	int i;
	for (i = 0; i < r->NPoints; i++) {
	    r->traceA[i] = MAX(0, r->traceA[i] - opts->subtract);
	    r->traceC[i] = MAX(0, r->traceC[i] - opts->subtract);
	    r->traceG[i] = MAX(0, r->traceG[i] - opts->subtract);
	    r->traceT[i] = MAX(0, r->traceT[i] - opts->subtract);
	}
    }

    if (opts->sub_background) { 
	/*
	trace_freq(r->traceA, r->NPoints);	
	trace_freq(r->traceC, r->NPoints);	
	trace_freq(r->traceG, r->NPoints);	
	trace_freq(r->traceT, r->NPoints);	
	*/
	subtract_background(r);
	/*
	separate_dyes(r, matrix);
	trace_freq(r->traceA, r->NPoints);	
	trace_freq(r->traceC, r->NPoints);	
	trace_freq(r->traceG, r->NPoints);	
	trace_freq(r->traceT, r->NPoints);
	*/
	reset_max_called_height(r);
    }

    if (opts->normalise) {
	rescale_heights(r, opts->min_normalise);
    }

    if (opts->scale) {
	rescale_trace(r, opts->scale);
    }

    if (opts->name)
	r->ident = strdup(opts->name);
    else if (0 == strcmp(outfname, "(stdout)"))
	r->ident = strdup(infname);
    else
	r->ident = strdup(outfname);

    if (opts->compress_mode != -1)
	set_compression_method(opts->compress_mode);

    if (0 != (mfwrite_reading(outfp, r, opts->out_format))) {
	fprintf(stderr, "failed to write file %s\n", outfname);
	read_deallocate(r);
	return 1;
    }

    read_deallocate(r);
    return 0;
}


void usage(void) {
    puts("Usage: convert_trace [options] [informat outformat] < in > out");
    puts("Or     convert_trace [options] -fofn file_of_filenames");
    puts("\nOptions are:");
    puts("\t-in_format format         Format for input (defaults to any");
    puts("\t-out_format format        Format for output (default ztr)");
    puts("\t-fofn file_of_filenames   Get \"Input Output\" names from a fofn");
    puts("\t-passed fofn              Output fofn of passed names");  
    puts("\t-failed fofn              Output fofn of failed names");  
    puts("\t-name id                  ID line for experiment file output");
    puts("\t-subtract_background      Auto-subtracts the trace background");
    puts("\t-subtract amount          Subtracts a specified background amount");
    puts("\t-normalise                Normalises peak heights");
    puts("\t-min_normalise            Minimum trace amp for normalising");
    puts("\t-scale range              Downscales peaks to 0-range");
    puts("\t-compress mode            Compress file output (not if stdout)");
    puts("\t-abi_data counts	      ABI DATA lanes to copy: eg 9,10,11,12");
    puts("\t--                        Explicitly state end of options");
    exit(1);
}

int main(int argc, char **argv) {
    struct opts opts;

    opts.in_format = TT_ANY;
    opts.out_format = TT_ZTR;
    opts.scale = 0;
    opts.sub_background = 0;
    opts.subtract = 0;
    opts.normalise = 0;
    opts.min_normalise = 100;
    opts.name = NULL;
    opts.compress_mode = -1;
    opts.dots = 0;
    opts.fofn = NULL;
    opts.passed = NULL;
    opts.failed = NULL;
    
    for (argc--, argv++; argc > 0; argc--, argv++) {
	if (**argv != '-')
	    break;

	if (strcmp(*argv, "-scale") == 0) {
	    opts.scale = atoi(*++argv);
	    argc--;

	} else if (strcmp(*argv, "-fofn") == 0) {
	    opts.fofn = *++argv;
	    argc--;

	} else if (strcmp(*argv, "-passed") == 0) {
	    opts.passed = *++argv;
	    argc--;

	} else if (strcmp(*argv, "-failed") == 0) {
	    opts.failed = *++argv;
	    argc--;

	} else if (strcmp(*argv, "-subtract_background") == 0) {
	    opts.sub_background = 1;

	} else if (strcmp(*argv, "-subtract") == 0) {
	    opts.subtract = atoi(*++argv);
	    argc--;

	} else if (strcmp(*argv, "-normalise") == 0) {
	    opts.normalise = 1;

	} else if (strcmp(*argv, "-min_normalise") == 0) {
	    opts.min_normalise = atoi(*++argv);
	    argc--;

	} else if (strcmp(*argv, "-dots") == 0) {
	    opts.dots = 1;

	} else if (strcmp(*argv, "-in_format") == 0) {
	    argv++;
	    argc--;
	    if (TT_UNK == (opts.in_format = trace_type_str2int(*argv)))
		opts.in_format = atoi(*argv);

	} else if (strcmp(*argv, "-name") == 0) {
	    opts.name = *++argv;
	    argc--;

	} else if (strcmp(*argv, "-out_format") == 0) {
	    argv++;
	    argc--;
	    if (TT_UNK == (opts.out_format = trace_type_str2int(*argv)))
		opts.out_format = atoi(*argv);

	} else if (strcasecmp(*argv, "-compress") == 0) {
	    opts.compress_mode = compress_str2int(*++argv);
	    argc--;

	} else if (strcmp(*argv, "-abi_data") == 0) {
	    int c1, c2, c3, c4;
	    argc--;
	    if (4 == sscanf(*++argv, "%d,%d,%d,%d", &c1, &c2, &c3, &c4)) {
		abi_set_data_counts(c1, c2, c3, c4);
	    } else {
		usage();
	    }

	} else if (strcmp(*argv, "--") == 0) {
	    break;

	} else {
	    usage();
	}
    }

    if (argc == 2) {
	/* Old syntax, for backwards compatibility */

	if (TT_UNK == (opts.in_format = trace_type_str2int(argv[0])))
	    opts.in_format = atoi(argv[0]);
	if (TT_UNK == (opts.out_format = trace_type_str2int(argv[1])))
	    opts.out_format = atoi(argv[1]);
    } else if (argc != 0) {
	usage();
    }

    if (!opts.fofn) {
	return convert(mstdin(), mstdout(), "(stdin)", "(stdout)", &opts);
    }

    /* else */ {
	mFILE *fpin, *fpout;
	FILE *fppassed = NULL, *fpfailed = NULL;
	char *infname, *outfname;
	int ret, ret_all = 0;
	char line[8192], line2[8192];

	FILE *fofn_fp;

	if (NULL == (fofn_fp = fopen(opts.fofn, "r"))) {
	    perror(opts.fofn);
	    return -1;
	}

	if (opts.passed && NULL == (fppassed = fopen(opts.passed, "w"))) {
	    perror(opts.passed);
	    return -1;
	}

	if (opts.failed && NULL == (fpfailed = fopen(opts.failed, "w"))) {
	    perror(opts.failed);
	    return -1;
	}

	while (fgets(line, 8192, fofn_fp) != NULL) {
	    int i, j, len;

	    /* Find input and output name, escaping spaces as needed */
	    len = strlen(line);
	    outfname = NULL;
	    for (i = j = 0; i < len; i++) {
		if (line[i] == '\\' && i != len-1) {
		    line2[j++] = line[++i];
		} else if (line[i] == ' ') {
		    line2[j++] = 0;
		    outfname = &line2[j];
		} else if (line[i] != '\n') {
		    line2[j++] = line[i];
		}
	    }
	    line2[j] = 0;
	    infname = line2;

	    /* Open input and output files */
	    if (opts.in_format == TT_EXP) {
		fpin = open_exp_mfile(infname, NULL);
	    } else {
		fpin = open_trace_mfile(infname, NULL);
	    }
	    if (NULL == fpin) {
		char buf[2048];
		sprintf(buf, "ERROR %s", infname);
		perror(buf);
		if (opts.dots) {
		    fputc('!', stdout);
		    fflush(stdout);
		}
		if (fpfailed)
		    fprintf(fpfailed, "%s\n", infname);
		continue;
	    }

	    if (outfname) {
		if (NULL == (fpout = mfopen(outfname, "wb+"))) {
		    char buf[2048];
		    sprintf(buf, "ERROR %s", outfname);
		    perror(buf);
		    mfclose(fpin);
		    if (opts.dots) {
			fputc('!', stdout);
			fflush(stdout);
		    }
		    if (fpfailed)
			fprintf(fpfailed, "%s\n", infname);
		    continue;
		}
	    } else {
		outfname = "(stdout)";
		fpout = mstdout();
	    }

	    /* Convert */
	    ret = convert(fpin, fpout, infname, outfname, &opts);
	    ret_all |= ret;
	    if (opts.dots) {
		fputc(ret ? '!' : '.', stdout);
		fflush(stdout);
	    }
	    if (ret) {
		if (fpfailed)
		    fprintf(fpfailed, "%s\n", infname);
	    } else {
		if (fppassed)
		    fprintf(fppassed, "%s\n", infname);
	    }

	    /* Tidy up */
	    mfclose(fpin);
	    if (fpout != mstdout())
		mfclose(fpout);
	}

	fclose(fofn_fp);

	return ret_all;
    }
}
